### ISSUE TITLE
Crash under load in mapnik::freetype_engine::create_face

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#2603]
More proof that `create_face` may not be fully thread safe.
- https://gist.github.com/springmeyer/6398dd7516009f36479d
- https://gist.github.com/springmeyer/eda2135cfecf0df8f1b3

refs #2198, #2209

**Comment by springmeyer** [comment]
@artemp yeah, I agree. To be 100% we'd ideally load fonts in advance to avoid needing to lock. So far this crash has been really really rare because once one thread succeeds fonts are cache and the crash can't happen. However, in the meantime I'm here to say that we found a case that that is more crashy: it was when fonts were being registered per tile load due to a logic bug in tilelive-mapnik: https://github.com/mapbox/tilelive-mapnik/commit/37a96814534910b4b7df48ce2fad119edd7defe4 which is related to https://github.com/mapnik/node-mapnik/issues/437#issuecomment-100390046

**Comment by flippmoke** [comment]
@springmeyer @artemp I think the safest and best thing might be to use a `boost::shared_mutex`. I think it would be performant enough for our needs and only would require locking prior to calling `create_face`. I do not believe obtaining a `shared_lock` is a very large hit on performance. Though we should do some benchmarks to test.

I honestly think that trying to load all fonts in advance is a nice solution, but might require more time at start up that simply isn't required in many operations.

The only downside is that this would require boost_thread in our core again, which I do not think it currently does?
